File: StairsGain.rb
def StairsGain s
  d, d[0], d[1] = [], 0, s[0] == 'y' ? 1 : -2

  (2..s.length).each do |i|
    t = s[i - 2] == 'y' ? 2 : - 1
    d[i] = s[i - 1] == 'y' ? [d[i - 1] + 1, d[i - 2] + t + 1].max : [d[i - 1] - 2, d[i - 2] + t - 2].max
  end
  d.last
end

File: codefights/lib/stairs_gain.ex
defmodule StairsGain do
  def stairsGain(s) do
    d = [0, (if s |> String.at(0) == "y", do: 1, else: -2)]
    Enum.reduce((2..(String.length(s))) |> Enum.to_list, d, fn(i, acc) ->
      t = if (s |> String.at(i - 2)) == "y", do: 2, else: -1
      acc ++ [(if (s |> String.at(i - 1)) == "y", do: Enum.max([Enum.at(acc, i - 1) + 1, Enum.at(acc, i - 2) + t + 1]), else: Enum.max([Enum.at(acc, i - 1) - 2, Enum.at(acc, i - 2) + t - 2]))]
    end) |> List.last
  end
end
